### PR TITLE
Add duplicate port validation, update devfile libraries, and add tests to check port name matches endpoint name

### LINKED ISSUE #4737: port names should match endpoint names
Assuming that devfile containers have following endpoints
```
      endpoints:
        - name: first
          targetPort: 3000
        - name: second
          targetPort: 4000
```

The k8s Service that gets generated for odo should have matching port names

Current situation:
```
  ports:
  - name: port-3000
    port: 3000
    protocol: TCP
    targetPort: 3000
  - name: port-4000
    port: 4000
    protocol: TCP
    targetPort: 4000
```

/priority high
/kind bug

### LINKED ISSUE #5387: Odd issue when using a name for a url creation when there is already a route in the devfile.yaml
/kind bug

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 
  2. Use the Google group if you have a question rather than a bug or feature request.

The group is at: https://groups.google.com/forum/#!forum/odo-users

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?

**Operating System:** Linux

**Output of `odo version`:** odo v2.5.0 (724f16e68)


## How did you run odo exactly?

```
git clone https://github.com/odo-devfiles/odo-quickstart
cd odo-quickstart/frontend
odo create nodejs frontend
odo url create nodejs --port 3000 --host foo.com
odo push
```


## Actual behavior

```sh
/odo-quickstart/frontend  main ✗                                                                                                                                                                                                                                                                                                                                   119d ◒  
▶ odo push

Validation
 ✓  Validating the devfile [59588ns]

Creating Services for component frontend
 ✓  Services are in sync with the cluster, no changes are required

Creating Kubernetes resources for component frontend
 ✓  Waiting for component to start [8s]
 ✓  Links are in sync with the cluster, no changes are required
 ✓  Waiting for component to start [7ms]
 ⚠  Unable to create ingress, missing host information for Endpoint http-3000, please check instructions on URL creation (refer `odo url create --help`)

```

Why does it show a warning for the ingress of http-3000? 

If I am creating a URL with the same port, why is it added to the devfile.yaml rather than overriding the current one in there?

See the devfile.yaml output:

```yaml
components:
- container:
    dedicatedPod: false
    endpoints:
    - name: http-3000
      secure: false
      targetPort: 3000
    - exposure: public
      name: nodejs
      path: /
      protocol: http
      secure: false
      targetPort: 3000
    image: registry.access.redhat.com/ubi8/nodejs-14:latest
    memoryLimit: 1024Mi
    mountSources: true
    sourceMapping: /project
  name: runtime
```

## Expected behavior

Expected behaviour is to for the one within the devfile.yaml to be overridden, or atleast a warning come up when you use `odo url create`... it's very odd that it will just append to the devfile.yaml and not check what's already in there.


## Any logs, error output, etc?

See above

### PR DESCRIPTION
**Pull request by valaparthvi**
**What type of PR is this:**

/kind bug

**What does this PR do / why we need it:**
This PR adds duplicate port validation, updates devfile libraries, and adds tests to ensure port name matches endpoint names.
**Which issue(s) this PR fixes:**

Fixes #4737

**PR acceptance criteria:**

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**
_Port name matches endpoint name validation_
```sh
odo create nodejs-vue myvue
odo url create myurl --port 3000 --host `minikube ip`
odo push
kubectl get svc myvue-app -oyaml -ojsonpath="{.spec.ports[*].name}"
```
_Duplicate port validation_
```sh
odo create nodejs-vue myvue
odo url create --port 8080
```

### CODE REVIEW

File: pkg/envinfo/url.go [hunk 1]
```diff
@@ -191,7 +191,11 @@ func (ei *EnvInfo) ValidateURL(url localConfigProvider.LocalURL) error {
 		}
 		for _, endpoint := range component.Container.Endpoints {
 			if endpoint.Name == url.Name && !ei.updateURL {
-				return fmt.Errorf("url %v already exist in devfile endpoint entry under container %v", url.Name, component.Name)
+				return fmt.Errorf("url %v already exists in devfile endpoint entry under container %q", url.Name, component.Name)
+			}
+			// Check for a duplicate port entry
+			if endpoint.TargetPort == url.Port {
```
Comment (feloy) [inline pkg/envinfo/url.go:197]: I'm not sure to understand when updateURL is set to true. Do you think it could be needed to test its value here also?

**Comment by feloy** [comment]
It seems you will have to adapt some tests that are using duplicated port numbers: https://s3.eu-de.cloud-object-storage.appdomain.cloud/odo-tests-openshift-logs/pr-5407-openshift-tests-441.html

File: pkg/envinfo/url.go (hunk 1 above)
Comment (valaparthvi) [reply pkg/envinfo/url.go]: I think it is set to true when you do not pass a URL name, and an invalid endpoint exists in the devfile. I'm not sure why it is as such, but that is my understanding.
https://github.com/redhat-developer/odo/blob/b75fa552ceb3daa2d37b4d931aca0f9b4686e86d/pkg/envinfo/url.go#L117-L131

**Comment by valaparthvi** [comment]
@feloy Please hold your review on this one for now.

**Comment by feloy** [comment]
This case is detected by tests on IBMCloud/Kubernetes but not on Prow tests:

```
â€¢ Failure [17.623 seconds]
odo devfile url command tests
/workspace/17f65e12-a2ec-43b6-bf47-4a8738e852ea/tests/integration/devfile/cmd_devfile_url_test.go:16
  when creating a Nodejs component
  /workspace/17f65e12-a2ec-43b6-bf47-4a8738e852ea/tests/integration/devfile/cmd_devfile_url_test.go:47
    should not allow to create URL with duplicate port [It]
    /workspace/17f65e12-a2ec-43b6-bf47-4a8738e852ea/tests/integration/devfile/cmd_devfile_url_test.go:70

    Expected
        <string>: I0209 06:58:40.232825   13236 util.go:816] Response will be cached in /tmp/odohttpcache for 1h0m0s
        I0209 06:58:40.236736   13236 util.go:829] Cached response used.
        I0209 06:58:40.405521   13236 implem.go:107] The path for preference file is /tmp/994218346/preference.yaml
        I0209 06:58:40.601851   13236 kclient.go:256] Checking if "routes" resource is supported
        I0209 06:58:40.607410   13236 context.go:141] absolute devfile path: '/tmp/730430071/devfile.yaml'
        I0209 06:58:40.607431   13236 context.go:92] absolute devfile path: '/tmp/730430071/devfile.yaml'
        I0209 06:58:40.608032   13236 content.go:33] converted devfile YAML to JSON
        I0209 06:58:40.608093   13236 apiVersion.go:45] devfile schemaVersion: '2.0.0'
        I0209 06:58:40.608110   13236 helper.go:41] devfile apiVersion '2.0.0' is supported
        I0209 06:58:40.686530   13236 schema.go:46] validated devfile schema
        I0209 06:58:40.703298   13236 validate.go:45] Successfully validated devfile sections
        I0209 06:58:40.705297   13236 validate.go:45] Successfully validated devfile sections
         âœ—  URL creation failed:
        	- host must be provided in order to create URLS of Ingress Kind
        
    to contain substring
        <string>: port 3000 already exists in devfile endpoint entry
```

**Comment by valaparthvi** [comment]
@feloy Can you review it again?

**Comment by cdrage** [comment]
Doesn't seem to work, or am I doing it incorrect?

```sh
git clone https://github.com/odo-devfiles/nodejs-ex
odo create nodejs foobar
odo push= 
odo url create --port 3000 foobar --host k8s.land
 ✗  URL creation failed:
        - port 3000 already exists in devfile endpoint entry under container "runtime"
 ```

**Comment by feloy** [comment]
> Doesn't seem to work, or am I doing it incorrect?
> 
> ```shell
> git clone https://github.com/odo-devfiles/nodejs-ex
> odo create nodejs foobar
> odo push=
> odo url create --port 3000 foobar --host k8s.land
>  ✗  URL creation failed:
>         - port 3000 already exists in devfile endpoint entry under container "runtime"
> ```

The error is expected, as the nodejs devfile already contains an endpoint for this specific port and you cannot create several endpoints with same port anymore:

```
$ odo create nodejs foobar
$ cat devfile.ymal
[...]
components:
- container:
    endpoints:
    - name: http-3000
      targetPort: 3000
    image: registry.access.redhat.com/ubi8/nodejs-14:latest
    memoryLimit: 1024Mi
    mountSources: true
    sourceMapping: /project
  name: runtime
[...]
```

**Comment by cdrage** [comment]
after discussion with @valaparthvi 

this

/approve

**Pull request merged**